### PR TITLE
initial podman implementation based on existing docker code.

### LINKED ISSUE #1397: podman support
## New feature

Starting with RHEL7.7 / RHEL8 RedHat is providing a new set of container tools like podman, buildah and skopeo.
Podman is almost a 1:1 replacement of docker commands and even an alias docker=podman does it's job. Nevertheless it would make sense to enable nextflow for podman.

## Usage scenario 

Run containers using podman.

## Suggest implementation 

Alias or duplicate the docker commands.

### PR DESCRIPTION
**Pull request by tbugfinder**
Signed-off-by: tbugfinder <tbugfinder@online.ms>

initial support of Podman.

fixes #1397

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -48,9 +48,9 @@
 # built documents.
 #
 # The short X.Y version.
-version = '19.12'
+version = '20.01'
 # The full version, including alpha/beta/rc tags.
-release = '19.12.0-edge'
+release = '20.01.0-SNAPSHOT'
```
Comment (pditommaso) [inline docs/conf.py]: Please do not commit changes to this file

File: modules/nextflow/src/main/groovy/nextflow/Const.groovy [hunk 2]
```diff
@@ -52,17 +52,17 @@ class Const {
...
      * The app build time as linux/unix timestamp
      */
-    static public final long APP_TIMESTAMP = 1576839997768
+    static public final long APP_TIMESTAMP = 1578594512367
 
     /**
      * The app build number
      */
-    static public final int APP_BUILDNUM = 5244
+    static public final int APP_BUILDNUM = 5250
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/Const.groovy]: Also these changes are generated automatically. Should not be committed.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    private static final String USER_AND_HOME_EMULATION = '-u $(id -u) -e "HOME=${HOME}" -v /etc/passwd:/etc/passwd:ro -v /etc/shadow:/etc/shadow:ro -v /etc/group:/etc/group:ro -v $HOME:$HOME'
+
+    private String removeCommand
+
+    private String killCommand
+
+    private kill = true
+
+    private boolean legacy
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: This flag is not needed.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+    private boolean userEmulation
+
+    private String registry
+
+    private String name
+
+    private boolean tty
+
+    private static final String USER_AND_HOME_EMULATION = '-u $(id -u) -e "HOME=${HOME}" -v /etc/passwd:/etc/passwd:ro -v /etc/shadow:/etc/shadow:ro -v /etc/group:/etc/group:ro -v $HOME:$HOME'
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: This option is not really used. Make sense remove from this implementation.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+ * @author tbugfinder <github@online.ms>
+ */
+@CompileStatic
+class PodmanBuilder extends ContainerBuilder<PodmanBuilder> {
+
+    private boolean sudo
+
+    private boolean remove = true
+
+    private boolean userEmulation
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: Not needed.

File: nextflow [hunk 6]
```diff
@@ -15,7 +15,7 @@
 #  limitations under the License.
 
 [[ "$NXF_DEBUG" == 'x' ]] && set -x
-NXF_VER=${NXF_VER:-'19.12.0-edge'}
+NXF_VER=${NXF_VER:-'20.01.0-SNAPSHOT'}
 NXF_ORG=${NXF_ORG:-'nextflow-io'}
```
Comment (pditommaso) [inline nextflow]: This is generated automatically. SHould not be included in the change set.

File: tests/checks/.IGNORE-PODMAN [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# TESTS TO BE IGNORED WHEN PODMAN
+# IS NOT AVAILABLE
+ampa.nf
+blast-parallel.nf
+blast.nf
+workdir-with-blank.nf
+stress.nf
+env-container.nf
```
Comment (pditommaso) [inline tests/checks/.IGNORE-PODMAN]: This can be removed.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+/**
+ * Helper methods to handle podman containers
+ *
+ * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
+ * @author tbugfinder <github@online.ms>
+ */
+@CompileStatic
+class PodmanBuilder extends ContainerBuilder<PodmanBuilder> {
+
+    private boolean sudo
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: The use of `sudo` as always been a bad practice with docker and with podman should not be needed in any case. Could this be removed ?

Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: One limit of rootless Podman is that cpuset cannot be applied. Therefore the possibility to still use Podman with sudo is included.
On the other hand Podman integration could be marked as <beta> until all features are supported.

File: docs/conf.py (hunk 1 above)
Comment (tbugfinder) [reply docs/conf.py]: removed from commit by
commit d94d05db3724e3e28605d2268193e209eb0d32d8

File: modules/nextflow/src/main/groovy/nextflow/Const.groovy (hunk 2 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/Const.groovy]: removed from commit by 
commit c8ce59e9d9ab182df7cd4077a255d3d50ba67f11

File: nextflow (hunk 6 above)
Comment (tbugfinder) [reply nextflow]: removed file from commit by
ommit 5a10e40292b8655136e9eca0abe632539af0cd13

File: tests/checks/.IGNORE-PODMAN (hunk 7 above)
Comment (tbugfinder) [reply tests/checks/.IGNORE-PODMAN]: removed file via commit fdae5e1fcc243962a79c284ff29d282c793349cd

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 3 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: Do you mean `kill` or about `legacy`?

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 4 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: please excuse, `tty` or `USER_AND_HOME_EMULATION`?

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 5 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: removed userEmulation with
commit 13b0fe2072ffb3202d3d0832ceebf4f991f7944a

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 4 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: removed USER_AND_HOME_EMULATION with
commit 13b0fe2072ffb3202d3d0832ceebf4f991f7944a

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 3 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: removed legacy flag with
commit 18eca201f4a77b8a9045f84a88fe13d7ae56be32

Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: Sorry, I mean `legacy`. This was related to old docker options.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 4 above)
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: Nice. Also `tty` can go, I think was never really used with docker.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 8 above)
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: I see, but it must be noted that cpuset is really a corner case and I don't think there's a concrete use in nextflow for it. Moreover, podman is useful because allows the user to run container in the user space without the need to escalate root. From this point of view makes no sense to give the `sudo` option. 

Maybe it could be useful instead to support `--cpus`. Does it require sudo as well?

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 4 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: removed tty with commit
commit cac990538e098d72152d2fbfdadd019c403b310a

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 8 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: `--cpus` can be used in rootless mode.
I agree on removing sudo.

Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: It could be part of engineOptions anyway.

Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: commit 3fb4e751c299b914e5f8f6c5d89e5e79a3fa63b2
Should the name be different than `cpus`?

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 4 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: removal of userEmulation was reverted
commit 3efc3b9c11c904512676cbb6f5f57ac7397c4243

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 8 above)
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: let's ignore it for now, NF is not to handle at this point

Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: Removed sudo, cpus and adopted the Tests.

Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: excellent

**Comment by pditommaso** [comment]
I have a question. Is podman able to automatically synchronise concurrent pulls of the same image?

**Comment by tbugfinder** [comment]
> I have a question. Is podman able to automatically synchronise concurrent pulls of the same image?

I've opened two shells each running the same pull & rm `$ podman pull godlovedc/lolcow ; Podman images ; podman image rm 577c1fe8e6d8` at the same time.
Both sessions pull the image but the first remove command removes the image for both (maybe obviously) - however the `rm` seems to wait for the other `pull` to complete. So the answer to your question is `no` with a bit of a `yes`.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 5 above)
Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: I'm sorry - I've added userEmulation and USER_AND_HOME_EMULATION again as I thought it's a useful shortcut. Please let me know if it's ok or should be finally removed.

**Comment by pditommaso** [comment]
Well, it looks more a yes. My concern is more about when launching two or more runs using the same image. Docker synchronise the pull, from what you are using podman should do the same.

File: modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy (hunk 5 above)
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: It was more an hack and with podman it should not be necessary. I would say to remove it.

Comment (tbugfinder) [reply modules/nextflow/src/main/groovy/nextflow/container/PodmanBuilder.groovy]: removed by 
commit 7dfa7c9169892505f45c3d42d13e83e2447f2ec9

**Comment by tbugfinder** [comment]
> Hello, thanks a lot for this PR. it 's very welcome. I've made a few comments on your submission.
> 
> Please also add an entry for the podman config in this snippet
> 
> https://github.com/nextflow-io/nextflow/blob/c477eeaff4762dcb409ed2eca954922ab1315ef2/modules/nextflow/src/main/groovy/nextflow/Session.groovy#L1092-L1095

this is already part of the PR.

**Comment by pditommaso** [comment]
OK, I think we can merge this. Thanks a lot for the contribution.
